fred2/missioncutscenesdlg.cpp

```diff
@@ -0,0 +1,466 @@
+{
+	m_cutscenes_tree.select_sexp_node = select_sexp_node;
+	select_sexp_node = -1;
+
+	m_cutscenes_tree.clear_tree();
+	m_cutscenes.clear();
+	m_sig.clear();
+	for (int i=0; i<(int)The_mission.cutscenes.size(); i++) {
+		m_cutscenes.push_back(The_mission.cutscenes[i]);
+		m_sig.push_back(i);
+
+		if (m_cutscenes[i].filename[0] == '\0')
+			strcpy_s(m_cutscenes[i].filename, "<Unnamed>");
+
+		m_cutscenes[i].formula = m_cutscenes_tree.load_sub_tree(The_mission.cutscenes[i].formula, true, "true");
+	}
+
+	m_cutscenes_tree.post_load();
+	cur_cutscene = -1;
+	update_cur_cutscene();
+}
+
+// create the CTreeCtrl tree from the cutscene tree, filtering based on m_display_cutscene_types
+void CMissionCutscenesDlg::create_tree()
+{
+	m_desc = _T(cutscene_descriptions[m_display_cutscene_types]);
+
+	m_cutscenes_tree.DeleteAllItems();
+	m_cutscenes_tree.reset_handles();
+	for (int i = 0; i < (int)m_cutscenes.size(); i++) {
+		if (m_cutscenes[i].type != m_display_cutscene_types)
+			continue;
+
+		HTREEITEM h = m_cutscenes_tree.insert(m_cutscenes[i].filename);
+		m_cutscenes_tree.SetItemData(h, m_cutscenes[i].formula);
+		m_cutscenes_tree.add_sub_tree(m_cutscenes[i].formula, h);
+	}
+
+	cur_cutscene = -1;
+	update_cur_cutscene();
+}
+
+// Display cutscene types selection changed, so update the display
+void CMissionCutscenesDlg::OnSelchangeDisplayCutsceneTypesDrop()
+{
+	UpdateData(TRUE);
+	create_tree();
+}
+
+// New tree item selected.  Because cutscene info is displayed for the selected tree item,
+// we need to update the display when this occurs.
+void CMissionCutscenesDlg::OnSelchangedCutscenesTree(NMHDR* pNMHDR, LRESULT* pResult)
+{
+	NM_TREEVIEW* pNMTreeView = (NM_TREEVIEW*)pNMHDR;
+
+	HTREEITEM h = pNMTreeView->itemNew.hItem;
+	if (!h)
+		return;
+
+	HTREEITEM h2;
+	m_cutscenes_tree.update_help(h);
+	while ((h2 = m_cutscenes_tree.GetParentItem(h)) != 0) {
+		h = h2;
+	}
+
+	int z = (int)m_cutscenes_tree.GetItemData(h);
+	int i;
+	for (i = 0; i < (int)m_cutscenes.size(); i++) {
+		if (m_cutscenes[i].formula == z) {
+			break;
+		}
+	}
+
+	Assert(i < (int)m_cutscenes.size());
+	cur_cutscene = i;
+	update_cur_cutscene();
+	*pResult = 0;
+}
+
+// update display info to reflect the currently selected cutscene.
+void CMissionCutscenesDlg::update_cur_cutscene()
+{
+	if (cur_cutscene < 0) {
+		m_name = _T("");
+		m_cutscene_type = -1;
+		UpdateData(FALSE);
+		GetDlgItem(IDC_CUTSCENE_TYPE_DROP) -> EnableWindow(FALSE);
+		GetDlgItem(IDC_CUTSCENE_NAME) -> EnableWindow(FALSE);
+
+		UpdateData(FALSE);
+		return;
+	}
+
+	m_name = _T(m_cutscenes[cur_cutscene].filename);
+	m_cutscene_type = m_cutscenes[cur_cutscene].type;
+
+	UpdateData(FALSE);
+	GetDlgItem(IDC_CUTSCENE_TYPE_DROP) -> EnableWindow(TRUE);
+	GetDlgItem(IDC_CUTSCENE_NAME) -> EnableWindow(TRUE);
+
+	UpdateData(FALSE);
+}
+
+// handler for context menu (i.e. a right mouse button click).
+void CMissionCutscenesDlg::OnRclickCutscenesTree(NMHDR* pNMHDR, LRESULT* pResult)
```

Comment:
Looks like pNMHDR is an unused parameter

Reply:
I tried removing it, but it seems like it's something that ON_NOTIFY expects to have. Build failed without it.

Reply:
Ok, no problem